sql/src/main/java/io/crate/jobs/PageDownstreamContext.java

```diff
@@ -51,10 +56,10 @@
     private final Streamer<?>[] streamers;
     private final RamAccountingContext ramAccountingContext;
     private final int numBuckets;
-    private final BitSet exhausted;
+    private final Set<Integer> exhausted;
     private final PagingIterator<Integer, Row> pagingIterator;
-    private final IntObjectHashMap<PageResultListener> listenersByBucketIdx;
-    private final IntObjectHashMap<Bucket> bucketsByIdx;
+    private final Map<Integer, PageResultListener> listenersByBucketIdx;
```

Comment:
Is the IntObjectHashMap -> Map change necessary? If I understand the problem correctly only the BitSet was problematic?
If the Map change is unrelated please revert it. It just hides the real fix and there is nothing wrong with the hppc map, quite the opposite, it avoids integer boxing in this case.

Reply:
I thought you would comment on this :)

> Is the IntObjectHashMap -> Map change necessary? 

Yes, because the iterator of `IntObjectHashMap` doesn't have a `remove()` method which we need to iterate and remove values. Granted, we could remove later on but that would require to store a temporary list of to-be-removed-items.

Strictly speaking, `bucketsByIdx` wouldn't have to be converted, but I rather use once consistent interface for storing the bucket ids, then two different interfaces and data structures.

> it avoids integer boxing in this case.

Not in all cases, since we have also callbacks with `Integer` into the hppc map. Anyhow, this is an extreme micro-optimization.

Reply:
> I thought you would comment on this 

yep because it's very hard to distinguish between relevant and unrelevant changes, and we generally have a tendency to sneak in unrelated changes. That makes reviews only more difficult.

> Yes, because the iterator of IntObjectHashMap doesn't have a remove() 

I think `remove(idx)` could be called even while the iterator is used.

> Anyhow, this is an extreme micro-optimization. 

I'm well aware, I wanted to make a point that it's not *worse* than a HashMap.

Anyway, I'll leave this up to you

Reply:
>yep because it's very hard to distinguish between relevant and unrelevant changes, and we generally have a tendency to sneak in unrelated changes. That makes reviews only more difficult.

I'm really against sneaking in anything and always try to keep my commits clean. However, sometimes it is better to change a bit more and sacrifice clarity over minimal invasive changes.

>I think remove(idx) could be called even while the iterator is used.

Honestly, I had it like that in the beginning but I wasn't sure if that would introduce a bug. I think the resizing on remove could break the iterator.

>I'm well aware, I wanted to make a point that it's not worse than a HashMap.

I don't understand why you would prefer a non-standard API if it is not really beneficial. It just makes it harder for people to read the code.